velox/exec/Merge.cpp

```diff
@@ -123,102 +104,321 @@ BlockingReason Merge::isBlocked(ContinueFuture* future) {
+  mergeOutputSpiller_ = nullptr;
+  VELOX_CHECK_EQ(spillPartitionSet.size(), 1);
+  spillFileGroups_.push_back(spillPartitionSet.begin()->second->files());
+  sourceMerger_ = nullptr;
   if (numStartedSources_ >= sources_.size()) {
+    setupSpillMerger();
+  }
+}
+
+void Merge::setupSpillMerger() {
+  VELOX_CHECK(!spillFileGroups_.empty());
+  VELOX_CHECK_NULL(spillMerger_);
+  std::vector<std::vector<std::unique_ptr<SpillReadFile>>> spillReadFileGroups;
+  spillReadFileGroups.reserve(spillFileGroups_.size());
+  for (const auto& spillFiles : spillFileGroups_) {
+    std::vector<std::unique_ptr<SpillReadFile>> spillReadFiles;
+    spillReadFiles.reserve(spillFiles.size());
+    for (const auto& spillFile : spillFiles) {
+      spillReadFiles.emplace_back(SpillReadFile::create(
+          spillFile, spillConfig_->readBufferSize, pool(), &spillStats_));
+    }
+    spillReadFileGroups.push_back(std::move(spillReadFiles));
+  }
+  spillFileGroups_.clear();
+  spillMerger_ = std::make_unique<SpillMerger>(
+      outputType_, numSpilledRows_, std::move(spillReadFileGroups), pool());
+}
+
+void Merge::maybeStartNextMergeSourceGroup() {
+  if (sourceMerger_ != nullptr || numStartedSources_ >= sources_.size()) {
     return;
   }
-  VELOX_CHECK_EQ(numStartedSources_, 0);
-  VELOX_CHECK(streams_.empty());
-  VELOX_CHECK(sourceBlockingFutures_.empty());
-  // TODO: support lazy start for local merge with a large number of sources
-  // to cap the memory usage.
-  for (auto& source : sources_) {
+
+  // Gets the merge sources for the next partial merge run.
+  std::vector<MergeSource*> sources;
+  for (auto i = numStartedSources_; i <
+       (std::min(sources_.size(), numStartedSources_ + maxNumMergeSources_));
+       ++i) {
+    sources.push_back(sources_[i].get());
+  }
+
+  // Initializes the source merger.
+  std::vector<std::unique_ptr<SourceStream>> cursors;
+  cursors.reserve(sources.size());
+  for (auto* source : sources) {
+    cursors.push_back(
+        std::make_unique<SourceStream>(source, sortingKeys_, outputBatchSize_));
+  }
+
+  sourceMerger_ =
+      std::make_unique<SourceMerger>(outputType_, std::move(cursors), pool());
+  // Start sources.
+  for (const auto& source : sources) {
     source->start();
   }
-  numStartedSources_ = sources_.size();
+  numStartedSources_ += sources.size();
 }
 
-bool Merge::isFinished() {
-  return finished_;
+RowVectorPtr Merge::getOutputFromSpill() {
+  VELOX_CHECK_NOT_NULL(spillMerger_);
+  VELOX_CHECK_NULL(sourceMerger_);
+  output_ = spillMerger_->getOutput(outputBatchSize_);
+  if (output_ != nullptr) {
+    return std::move(output_);
+  }
+  spillMerger_.reset();
+  finished_ = true;
+  return nullptr;
+}
+
+RowVectorPtr Merge::getOutputFromSource() {
+  VELOX_CHECK_NULL(spillMerger_);
+  bool atEnd = false;
+  output_ =
+      sourceMerger_->getOutput(outputBatchSize_, sourceBlockingFutures_, atEnd);
+
+  if (needSpill()) {
+    spill();
+    if (atEnd) {
+      finishMergeSourceGroup();
```

Comment:
Let's call this right after sourceMerger_->getOutput(outputBatchSize_, sourceBlockingFutures_, atEnd), and we check internally if we need to finish spill or not? But it is always good to reset the source merger? Thanks!